File: packages/caelus/src/ranges.ts
/**
 * astroengine validated ranges -- the measured spans behind the headline, and
 * the delta-T uncertainty that bounds what any ephemeris can say about ancient
 * dates.
 *
 * The engine's analytic series (VSOP87D, Meeus Moon/Pluto, the Kepler pack)
 * return numbers at any epoch; what is *validated* is narrower and per body.
 * This module is the one source of truth for those spans (mirroring
 * `docs/range-expansion.md` and `accuracy.json` -- "validated, not asserted":
 * the figures move only when the Horizons measurement does), consumed by
 * {@link engineCapabilities} for introspection and by chart assembly for
 * per-chart warnings.
 *
 * It is a leaf module (imports only core types) so both `chart.ts` and
 * `capabilities.ts` can use it without an import cycle.
 */
import type { EngineData, ChebData } from "./core.js";

/** A calendar-year span, inclusive. */
export interface BodySpan {
  from: number;
  to: number;
}

/** The measured headline band. The label is registry-pinned to accuracy.json:
 *  1850-2150. */
export const HEADLINE: BodySpan = { from: 1850, to: 2150 };
export const HEADLINE_LABEL = "1850-2150";

/** Measured spans that are narrower or wider than the headline
 *  (docs/range-expansion.md's table). */
export const MEASURED: Record<string, BodySpan> = {
  pluto_pack: { from: 1700, to: 2212 },
  pluto_meeus: { from: 1885, to: 2099 },
  uranian: { from: 1800, to: 2149 },
};

export const VSOP_BODIES = new Set([
  "sun", "mercury", "venus", "mars", "jupiter", "saturn", "uranus", "neptune",
]);

/** Analytic points with no stated range bound (docs/range-expansion.md). */
export const ANALYTIC_BODIES = new Set([
  "mean_node", "true_node", "mean_lilith", "true_lilith",
]);

/** Approximate calendar year for a Julian Day (for reporting spans). */
export function jdYear(jd: number): number {
  return 2000 + (jd - 2451545.0) / 365.25;
}

/** A pack's fitted span in calendar years (`jd0 + segments * seg_days`). */
export function packSpan(pack: ChebData): BodySpan {
  return {
    from: Math.ceil(jdYear(pack.jd0)),
    to: Math.floor(jdYear(pack.jd0 + pack.segments.length * pack.seg_days)),
  };
}

/**
 * The measured validated span for a body under the given engine data, or
 * `null` for analytic points with no stated bound and unknown (synthetic)
 * bodies. Packed bodies report the headline: outside their *fitted* span
 * they are not computed at all (they land in `Chart.unavailable`), so the
 * validated question only arises inside it.
 */
export function validatedSpanFor(body: string, data: EngineData): BodySpan | null {
  // intp_apog falls through to null deliberately: a derived point (its own
  // EngineData field, not a chebPacks entry) with no measured span yet --
  // validate_swiss.py measures it against SE_INTP_APOG where pyswisseph
  // exists, and degree-scale differences are expected by construction.
  if (VSOP_BODIES.has(body) || body === "moon") return HEADLINE;
  if (body === "pluto") {
    return data.chebPacks?.pluto ? MEASURED.pluto_pack : MEASURED.pluto_meeus;
  }
  if (body === "chiron" || body in (data.chebPacks ?? {})) return HEADLINE;
  if (body in (data.keplerPack?.bodies ?? {})) return MEASURED.uranian;
  return null;
}

/**
 * One-sigma uncertainty of delta T (TT - UT1) in seconds at a calendar year.
 *
 * Breakpoints from Morrison & Stephenson, "Historical values of the Earth's
 * clock error deltaT and the calculation of eclipses" (J. Hist. Astron. 35,
 * 2004), Table 1's sigma column; the telescopic era tapers to the IERS
 * measured span (effectively exact), and the future grows quadratically from
 * the end of the measured tables (extrapolation, stated as such). Linear
 * interpolation between breakpoints.
 *
 * This is what bounds ancient charts: at year 150 the position theories are
 * fine but sigma(deltaT) is minutes, which smears the fast movers -- the
 * Ascendant/MC rotate 0.25 deg per minute of clock error and the Moon moves
 * 0.55 arcsec per second -- while Jupiter barely notices. Chart warnings
 * report exactly that split.
 */
export function deltaTSigma(year: number): number {
  // [year, sigma seconds]
  const TABLE: [number, number][] = [
    [-1000, 640], [-500, 430], [0, 260], [500, 160], [1000, 55],
    [1200, 30], [1400, 20], [1600, 20], [1700, 5], [1750, 2],
    [1800, 1], [1850, 0.5], [1900, 0.1], [2025, 0.05],
  ];
  if (year <= TABLE[0][0]) return TABLE[0][1];
  const last = TABLE[TABLE.length - 1];
  if (year > last[0]) {
    // Future extrapolation: the long-term deltaT forecasts spread by a few
    // seconds per generation; grow quadratically from the measured end.
    const t = (year - last[0]) / 50;
    return last[1] + 2.0 * t * t;
  }
  for (let i = 1; i < TABLE.length; i++) {
    const [y1, s1] = TABLE[i];
    if (year <= y1) {
      const [y0, s0] = TABLE[i - 1];
      return s0 + (s1 - s0) * ((year - y0) / (y1 - y0));
    }
  }
  return last[1];
}

File: python/astroengine/ranges.py
"""Validated ranges + delta-T sigma -- mirror of packages/caelus/src/ranges.ts.

One source of truth for the measured per-body spans behind the headline
("validated, not asserted": the figures move only when the Horizons
measurement does) and the Morrison & Stephenson (2004) delta-T uncertainty,
consumed by chart warnings so both engines emit identical validity
statements (the canonical digest pins them equal cross-language).
"""
from .core import delta_t_sigma, _moon_era_slabs  # re-export site for symmetry with ranges.ts

HEADLINE = {"from": 1000, "to": 3000}  # mirror of ranges.ts; see its comment

MEASURED = {
    "pluto_pack": {"from": 1000, "to": 3000},
    "pluto_meeus": {"from": 1885, "to": 2099},
    "uranian": {"from": 1800, "to": 2149},
    # Small bodies are fitted/validated only where Horizons serves them
    # (~1600-2500); narrower than the headline (see ranges.ts).
    "small_bodies": {"from": 1600, "to": 2484},
}

VSOP_BODIES = {"sun", "mercury", "venus", "mars", "jupiter", "saturn",
               "uranus", "neptune"}

ANALYTIC_BODIES = {"mean_node", "true_node", "mean_lilith", "true_lilith"}


def jd_year(jd):
    """Approximate calendar year for a Julian Day."""
    return 2000 + (jd - 2451545.0) / 365.25


# Mirror of ranges.ts: the two clock-error rates and the display threshold
# behind the canonical epoch-sigma block.
ANGLE_SIGMA_DEG_PER_SECOND = 0.25 / 60   # sky rotation: 0.25 deg per minute
MOON_SIGMA_DEG_PER_SECOND = 0.55 / 3600  # Moon longitude: ~0.55 arcsec per second
EPOCH_SIGMA_DISPLAY_SECONDS = 10         # below this, canonical output carries no block


def pack_span(pack):
    """A pack's fitted span in calendar years (mirror of packSpan in
    ranges.ts)."""
    import math
    return {"from": math.ceil(jd_year(pack.jd0)),
            "to": math.floor(jd_year(pack.jd1))}


def validated_span_for(body, engine):
    """Measured validated span for a body under this engine's data, or None
    for analytic points with no stated bound (mirrors validatedSpanFor;
    intp_apog falls through to None deliberately -- a derived point pending
    the SE measurement)."""
    from .chart import ASTEROIDS, URANIANS
    if body in VSOP_BODIES or body == "moon":
        span = HEADLINE
    elif body == "pluto":
        span = MEASURED["pluto_pack"] if engine._has_pluto_pack() \
            else MEASURED["pluto_meeus"]
    elif body == "chiron" or body in ASTEROIDS:
        span = MEASURED["small_bodies"]
    elif body in URANIANS:
        span = MEASURED["uranian"]
    else:
        span = None
    # Era slabs extend the validated span (deep-time R1): a slab's span is
    # measured against its source at pack build, and the classical slab
    # adjoins the headline's earlier edge. The Moon's slabs live in their own
    # loader beside its tiered packs.
    slabs = (_moon_era_slabs() if body == "moon"
             else engine._era_slabs(body))
    if slabs:
        frm = min(pack_span(s)["from"] for s in slabs)
        to = max(pack_span(s)["to"] for s in slabs)
        span = ({"from": min(span["from"], frm), "to": max(span["to"], to)}
                if span else {"from": frm, "to": to})
    return span


def _fmt(x):
    """Render a number the way a JS template literal does (2 not 2.0)."""
    if isinstance(x, float) and x.is_integer():
        return str(int(x))
    return str(x)


def chart_warnings(jd_ut, present_bodies, engine):
    """Validity statements for a chart -- mirrors chartWarnings in chart.ts,
    text for text (the canonical digest depends on byte equality)."""
    year = jd_year(jd_ut)
    warnings = []
    for body in present_bodies:
        span = validated_span_for(body, engine)
        if span and (year < span["from"] or year > span["to"]):
            warnings.append({
                "kind": "outside_validated_range", "body": body,
                "validated": span,
                "text": (f"{body} is outside its validated range "
                         f"({span['from']}-{span['to']}) at this instant: the "
                         "position is computed but not validated here"),
            })
    sigma = delta_t_sigma(year)
    if sigma >= 10:
        angle_smear = round_2(sigma / 60 * 0.25)
        moon_smear = round_2(sigma * 0.55 / 60)
        warnings.append({
            "kind": "delta_t_uncertain",
            "sigmaSeconds": js_round(sigma),
            "angleSmearDeg": angle_smear,
            "moonSmearArcmin": moon_smear,
            "text": (f"delta-T is uncertain by ~{js_round(sigma)} s at this "
                     "epoch (Morrison & Stephenson 2004): the Ascendant/MC "
                     f"smear ~{_fmt(angle_smear)} deg and the Moon "
                     f"~{_fmt(moon_smear)} arcmin, while the slow bodies are "
                     "unaffected"),
        })
    return warnings


def js_round(x):
    """JS Math.round (half toward +infinity), IEEE-guarded."""
    import math
    r = math.floor(x + 0.5)
    if r - x > 0.5:
        return int(r - 1)
    return int(r)


def round_2(x):
    """JS Math.round(x * 100) / 100."""
    return js_round(x * 100) / 100
